Example/GiniVision/ComponentAPICoordinator.swift

```diff
@@ -0,0 +1,269 @@
+            return
+        }
+        
+        // Restart analysis if it was canceled and is currently not running.
+        if !AnalysisManager.sharedManager.isAnalyzing {
+            if let documentData = document?.data {
+                AnalysisManager.sharedManager.analyzeDocument(withData: documentData, cancelationToken: CancelationToken(), completion: nil)
+            }
+        }
+        
+        // Show analysis screen if no results are in yet and no changes were made.
+        showAnalysisScreen(withDocument: documentReviewed)
+    }
+    
+    func didCancelReview() {
+        AnalysisManager.sharedManager.cancelAnalysis()
+    }
+}
+
+// MARK: ComponentAPIAnalysisScreenDelegate
+
+extension ComponentAPICoordinator: ComponentAPIAnalysisScreenDelegate {
+    func didTapErrorButton() {
+        if let document = document {
+            AnalysisManager.sharedManager.analyzeDocument(withData: document.data, cancelationToken: CancelationToken(), completion: nil)
+        }
+    }
+    
+    func didCancelAnalysis() {
+        AnalysisManager.sharedManager.cancelAnalysis()
+    }
+    
+    func didAppear() {
+        // Subscribe to analysis events which will be fired when the analysis process ends.
+        // Either with a valid result or an error.
+        NotificationCenter.default.addObserver(self, selector: #selector(handleAnalysis(errorNotification:)), name: NSNotification.Name(rawValue: GINIAnalysisManagerDidReceiveErrorNotification), object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(handleAnalysis(resultNotification:)), name: NSNotification.Name(rawValue: GINIAnalysisManagerDidReceiveResultNotification), object: nil)
+        
+        // Because results may come in during view controller transition,
+        // check for already existent results in shared analysis manager.
+        handleExistinResults()
+    }
+    
+    func didDisappear() {
+        NotificationCenter.default.removeObserver(self)
+    }
+
+}
+
+// MARK: NoResultsScreenDelegate
+
+extension ComponentAPICoordinator: NoResultsScreenDelegate {
+    func didTapRetry() {
+        if let navVC = navigationController, navVC.viewControllers.count != 1 {
+            _ = navVC.popToRootViewController(animated: true)
+        } else {
+            dismissTabBarController()
+        }
+    }
+}
+
+// MARK: Handle analysis results
+
+extension ComponentAPICoordinator {
+    fileprivate func handleAnalysis(_ result: GINIResult, fromDocument document: GINIDocument) {
+        let payFive = ["paymentReference", "iban", "bic", "paymentReference", "amountToPay"]
+        let hasPayFive = result.filter { payFive.contains($0.0) }.count > 0
+        
+        if hasPayFive {
+            showResultsTableScreen(forDocument: document, withResult: result)
+        } else {
+            showNoResultsScreen()
+        }
+        
+        if let analysisScreen = analysisScreen {
+            removeFromStack(analysisScreen)
```

Comment:
Wont it always be the top view controller?

Reply:
I think that it was made thinking only on the original flow, but as I said before, this method will be removed in a future refactor.